File: templates/r/rcs.R
## 限制性立方样条 RCS 剂量反应（非线性 OR + 95%CI 带）—— 卡片 13_因果推断图/限制性立方样条_RCS.md
## base splines::ns + glm（免 rms），相对参考点求 OR。 Rscript templates/r/rcs.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
suppressPackageStartupMessages(library(splines))
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_rcs"; TITLE <- "RCS dose-response"
XLAB <- "Exposure"; YLAB <- "Odds ratio (95% CI)"
KNOT_Q <- c(.05, .35, .65, .95)       # 4 节点位置（Harrell 推荐分位）
LINE_C <- unname(pal("med_case_control")["low"]); BAND_C <- LINE_C
# 换真实数据： x <- df$exposure; y <- df$outcome（0/1）
n <- 1500; x <- rnorm(n)
y <- rbinom(n, 1, plogis(0.4 * x^2 - 0.3 * x - 0.5))   # J 形真实关系
REF <- median(x)                      # 参考点（OR=1）
# <<< PARAM ------------------------------------------------------

knots <- quantile(x, KNOT_Q); bk <- knots[c(1, length(knots))]; ik <- knots[-c(1, length(knots))]
fit <- glm(y ~ ns(x, knots = ik, Boundary.knots = bk), family = binomial)
grid <- seq(quantile(x, .01), quantile(x, .99), length.out = 200)
mm  <- model.matrix(~ ns(grid, knots = ik, Boundary.knots = bk))
mmr <- model.matrix(~ ns(REF,  knots = ik, Boundary.knots = bk))
ct <- sweep(mm, 2, mmr, "-")          # 相对参考点的对比（截距相消）
lp <- as.vector(ct %*% coef(fit))
se <- sqrt(rowSums((ct %*% vcov(fit)) * ct))
d <- data.frame(x = grid, OR = exp(lp), lo = exp(lp - 1.96 * se), hi = exp(lp + 1.96 * se))
p <- ggplot(d, aes(x, OR)) +
  geom_ribbon(aes(ymin = lo, ymax = hi), fill = BAND_C, alpha = .18) +
  geom_line(color = LINE_C, linewidth = .9) +
  geom_hline(yintercept = 1, linetype = 2, color = "grey50") +
  geom_vline(xintercept = REF, linetype = 3, color = "grey50") +
  labs(x = XLAB, y = YLAB, title = TITLE)
save_gg(p, NAME, w = 100, h = 82)

File: templates/python/rcs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""限制性立方样条 RCS 剂量反应曲线（非线性 OR + 95%CI 带）—— 卡片 13_因果推断图/限制性立方样条_RCS.md
合成 J 形暴露-结局，RCS(4 节点) logistic 拟合，参考点 OR=1。改 # >>> PARAM 区。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, matplotlib.pyplot as plt, statsmodels.api as sm
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_rcs"
TITLE = "RCS dose-response"; XLAB = "Exposure"; YLAB = "Odds ratio (95% CI)"
KNOT_Q = [.05, .35, .65, .95]         # 4 节点位置（Harrell 推荐分位）
LINE_C = pal("med_case_control")["low"]; BAND_C = LINE_C
FIGSIZE = (4.0, 3.0)
# 换真实数据： x = df["exposure"].values; y = df["outcome"].values（0/1）
n = 1500; x = np.random.normal(0, 1, n)
y = np.random.binomial(1, 1 / (1 + np.exp(-(0.4 * x ** 2 - 0.3 * x - 0.5))))  # J 形真实关系
REF = float(np.median(x))             # 参考点（OR=1）
# <<< PARAM ------------------------------------------------------

def rcs_basis(v, knots):              # Harrell 限制性立方样条基（k 节点 → k-1 列）
    v = np.asarray(v, float); t = np.asarray(knots, float); k = len(t)
    cube = lambda u: np.where(u > 0, u ** 3, 0.0)
    cols = [v]; tk, tk1 = t[-1], t[-2]
    for j in range(k - 2):
        cols.append(cube(v - t[j]) - cube(v - tk1) * (tk - t[j]) / (tk - tk1)
                    + cube(v - tk) * (tk1 - t[j]) / (tk - tk1))
    return np.column_stack(cols)

knots = np.quantile(x, KNOT_Q)
fit = sm.GLM(y, sm.add_constant(rcs_basis(x, knots)), family=sm.families.Binomial()).fit()
grid = np.linspace(np.percentile(x, 1), np.percentile(x, 99), 200)
Bg = sm.add_constant(rcs_basis(grid, knots), has_constant="add")
Br = sm.add_constant(rcs_basis([REF], knots), has_constant="add")
contrast = Bg - Br                    # 相对参考点的对比（截距项相消）
lp = contrast @ fit.params
se = np.sqrt(np.einsum("ij,jk,ik->i", contrast, fit.cov_params(), contrast))
OR, lo, hi = np.exp(lp), np.exp(lp - 1.96 * se), np.exp(lp + 1.96 * se)

fig, ax = plt.subplots(figsize=FIGSIZE)
ax.fill_between(grid, lo, hi, color=BAND_C, alpha=.18)
ax.plot(grid, OR, color=LINE_C, lw=1.8)
ax.axhline(1, ls="--", color="grey", lw=.8)
ax.axvline(REF, ls=":", color="grey", lw=.8)
ax.set_xlabel(XLAB); ax.set_ylabel(YLAB); ax.set_title(TITLE); ax.grid(False)
save_fig(fig, NAME)
